firstapp/app/src/androidTest/java/com/example/firstapp/ShellUtility.java

```diff
@@ -67,17 +82,24 @@ UiObject executeUncachedAction() throws IOException, InterruptedException, UiObj
             }
             UiObject object = castToObject(object2);
             object.click();
-            return object;
+            cachedObject = object;
+        }
+        long executeCachedAction() throws UiObjectNotFoundException {
+            cachedObject.waitForExists(timeoutMs);
```

Comment:
is your idea here that the user would be able to specify timeoutMs?

Reply:
The idea is to make each Action more testable.  So less dependencies on the enclosing class the better.  Would be best to make the class "static", which decouples the nested class from the enclosing class.

Reply:
is it reasonable to leave the dependency on the device? I don't see any purpose in testing with a device explicitly, since the ShellUtility constructor identifies the connected device anyways

Reply:
I think it is okay now.  It was more of a wishlist.